src/components/MultiGestureCanvas/usePanGesture.js

```diff
@@ -0,0 +1,241 @@
+
+const PAN_DECAY_DECELARATION = 0.9915;
+
+const clamp = MultiGestureCanvasUtils.clamp;
+const SPRING_CONFIG = MultiGestureCanvasUtils.SPRING_CONFIG;
+
+const usePanGesture = ({
+    canvasSize,
+    contentSize,
+    panGestureRef,
+    pagerRef,
+    singleTap,
+    doubleTap,
+    zoomScale,
+    zoomRange,
+    totalScale,
+    totalOffsetX,
+    totalOffsetY,
+    panTranslateX,
+    panTranslateY,
+    isSwipingVertically,
+    isSwipingHorizontally,
+    onSwipeSuccess,
+    stopAnimation,
+}) => {
+    const zoomScaledContentWidth = useDerivedValue(() => contentSize.width * totalScale.value, [contentSize.width]);
+    const zoomScaledContentHeight = useDerivedValue(() => contentSize.height * totalScale.value, [contentSize.height]);
+
+    const previousTouch = useSharedValue(null);
+    // pan velocity to calculate the decay
+    const panVelocityX = useSharedValue(0);
+    const panVelocityY = useSharedValue(0);
+    // disable pan vertically when content is smaller than screen
+    const canPanVertically = useDerivedValue(() => canvasSize.height < zoomScaledContentHeight.value, [canvasSize.height]);
+
+    // calculates bounds of the scaled content
+    // can we pan left/right/up/down
+    // can be used to limit gesture or implementing tension effect
+    const getBounds = useWorkletCallback(() => {
+        let rightBoundary = 0;
+        let topBoundary = 0;
+
+        if (canvasSize.width < zoomScaledContentWidth.value) {
+            rightBoundary = Math.abs(canvasSize.width - zoomScaledContentWidth.value) / 2;
+        }
+
+        if (canvasSize.height < zoomScaledContentHeight.value) {
+            topBoundary = Math.abs(zoomScaledContentHeight.value - canvasSize.height) / 2;
+        }
+
+        const maxVector = {x: rightBoundary, y: topBoundary};
+        const minVector = {x: -rightBoundary, y: -topBoundary};
+
+        const target = {
+            x: clamp(totalOffsetX.value, minVector.x, maxVector.x),
+            y: clamp(totalOffsetY.value, minVector.y, maxVector.y),
+        };
+
+        const isInBoundaryX = target.x === totalOffsetX.value;
+        const isInBoundaryY = target.y === totalOffsetY.value;
+
+        return {
+            target,
+            isInBoundaryX,
+            isInBoundaryY,
+            minVector,
+            maxVector,
+            canPanLeft: target.x < maxVector.x,
+            canPanRight: target.x > minVector.x,
+        };
+    }, [canvasSize.width, canvasSize.height]);
+
+    const returnToBoundaries = useWorkletCallback(() => {
+        const {target, isInBoundaryX, isInBoundaryY, minVector, maxVector} = getBounds();
+
+        if (zoomScale.value === zoomRange.min && totalOffsetX.value === 0 && totalOffsetY.value === 0 && panTranslateX.value === 0 && panTranslateY.value === 0) {
+            // we don't need to run any animations
+            return;
+        }
+
+        if (zoomScale.value <= zoomRange.min) {
+            // just center it
```

Comment:
Please improve comment to explain why this case is necessary. Something like:

> When the user tries to zoom closer than the minimum zoom range, the subject will only be centered to prevent more zoom.